### PR TITLE
Add clang-format (pre-commit) and unit test (pre-push)  githooks 

### PR DESCRIPTION
**Pull request by chutsu**
In relation to #130 

To expedite our development process we use a set of git hooks to run:

- Clang format (pre-commit)
- Unit tests (pre-push)

these can be activated by running the `githook_init.bash` from the root of the repo

    bash scripts/githooks/githooks_init.bash

#### Pre-Merge Checklist:
- [x] Code is documented in doxygen format
- [x] Code has automated tests
- [x] Zero compiler warnings
- [x] Formatted with `clang-format`

### CODE REVIEW

File: scripts/githooks/pre-commit [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
```
Comment (leokoppel) [inline scripts/githooks/pre-commit]: Can this hook run `scripts/format.bash`?
That is what Travis runs to check the formatting, so there will be automatically be consistency between the formatting script and the check.

File: scripts/githooks/pre-push [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+mkdir -p build
+cd build
+cmake ..
+make -j`nproc --all`
+ctest
```
Comment (leokoppel) [inline scripts/githooks/pre-push]: This seems a little overzealous. For example, I usually run tests in my IDE and don't want to recompile on each push, or I push after a trivial documentation change. I may even want to push to a topic branch despite failing tests (e.g. halfway through implementing a feature).

File: scripts/githooks/pre-commit (hunk 1 above)
Comment (chutsu) [inline scripts/githooks/pre-commit]: Do you mind if I make format work in 2 different modes? At current "blanket" mode is by default, I would like to add the ability to only clang-format individual files.

Comment (leokoppel) [inline scripts/githooks/pre-commit]: You can already give a list of paths to format.bash; I put that in at your request : )

File: scripts/githooks/pre-push (hunk 2 above)
Comment (chutsu) [inline scripts/githooks/pre-push]: Removed

File: scripts/githooks/pre-commit (hunk 1 above)
Comment (chutsu) [inline scripts/githooks/pre-commit]: Done

File: README.md [hunk 3]
```diff
@@ -43,6 +43,19 @@ describing concepts and theory behind the library.
 
 It is generated using Doxygen, using the build target `make doc`.
 
+
+## Notes for Developers
+
+To expedite our development process we use a set of git hooks to run:
+
+- Clang format (pre-commit)
+- Unit tests (pre-push)
```
Comment (leokoppel) [inline README.md]: just update this

Comment (chutsu) [inline README.md]: Done

**Pull request merged**